README.md
xmos_tdm
========

Combined repository for a Clean XMOS AVB build

This repository is a combination of source code provided by XMOS on github, mainly sw_avb_lc, sc_avb, sc_ethernet, sc_otp_board_info, sc_util, sc_slicekit_support and sc_i2s.

The only additional stuff added is TDM support provided by Thomas Gmeinder also on github.

This is not a official XMOS release as it consists out of the most current github releases for each of the listed XMOS repositories.


CHANGELOG.rst
<<<<<<< HEAD
sc_slicekit_support Change Log
==============================

1.0.4
-----
  * Fix to the metainfo.

1.0.3
-----
  * Moved from sc_util
  * Fix code that assumes tile[0] gets node ide 0

1.0.2
-----
  * Fix module_slicekit_support to work with L16 target

1.0.1
-----
  * No change to slicekit support (changes to sc_util where module
    used to be)

1.0.0
-----
  * Initial Version
=======
sc_i2c Change Log
=================

2.4.1
-----
  * module_i2c_simple header-file comments updated to correctly reflect API

2.4.0
-----
  * i2c_shared functions now take i2cPorts structure as param (rather than externed). This allows for
    multiple i2c buses

  * Changes to dependencies:

    - sc_util: 1.0.3rc0 -> 1.0.4rc0

      + module_logging now compiled at -Os
      + debug_printf in module_logging uses a buffer to deliver messages unfragmented
      + Fix thread local storage calculation bug in libtrycatch
      + Fix debug_printf itoa to work for unsigned values > 0x80000000

2.3.0
-----
  * module_i2c_simple fixed to ACK correctly during multi-byte reads (all but the final byte will be now be ACKd)
  * module_i2c_simple can now be built with support to send repeated starts and retry reads and writes NACKd by slave
  * module_i2c_shared added to allow multiple logical cores to safely share a single I2C bus
  * Removed readreg() function from single_port module since it was not safe

2.2.1
-----
  * Documentation fixes

2.2.0
-----
  * Added I2C read/write functions with support of 16bit addresses
  * reg_read and reg_write now use nbytes params to allow multiple byte register access
  * Added support for i2c simple SDA on a 4 bit port
  * data[0] is no longer overwritten in module_i2c_simple read

2.1.0
-----
  * Updated documents for xSOFTip requirements
  * Added metainfo and XPD items

2.0.0
-----
  * Initial Version
  * Previous version on xmos.com
>>>>>>> a1891053e0f102301d8c0c64f3e4dd998681c0df


app_avb_lc_demo/src/avb_conf.h
#ifndef __avb_conf_h__
#define __avb_conf_h__

#include "app_config.h"

#if(AVB_AUDIO_IF_i2s)
#define AVB_AUDIO_IF_SAMPLES_PER_PERIOD 2
#endif
#if(AVB_AUDIO_IF_tdm_multi)
#define AVB_AUDIO_IF_SAMPLES_PER_PERIOD 8
#endif

/* Some of the configuration depends on the app_config.h file included above */

/******** ETHERNET MAC CONFIGURATION PARAMETERS *************************************************/
#define ETHERNET_DEFAULT_IMPLEMENTATION full

#define MAX_ETHERNET_PACKET_SIZE (1518)
#define NUM_MII_RX_BUF 6
#define NUM_MII_TX_BUF 3
#define ETHERNET_RX_HP_QUEUE 1
#define ETHERNET_TX_HP_QUEUE 1
#define MAX_ETHERNET_CLIENTS   4

#define ETHERNET_MAX_TX_HP_PACKET_SIZE (100 + AVB_DEMO_NUM_CHANNELS * 24)
#define ETHERNET_MAX_TX_LP_PACKET_SIZE (600)

#define MII_RX_BUFSIZE_HIGH_PRIORITY (1100 + (3*(ETHERNET_MAX_TX_HP_PACKET_SIZE)))
#define MII_RX_BUFSIZE_LOW_PRIORITY (2000)

#define MII_TX_BUFSIZE_HIGH_PRIORITY (1100 + (3*(ETHERNET_MAX_TX_HP_PACKET_SIZE)))
#define MII_TX_BUFSIZE_LOW_PRIORITY (2000)

#define NUM_ETHERNET_PORTS 1
#define NUM_ETHERNET_MASTER_PORTS 1

#define ETHERNET_RX_ENABLE_TIMER_OFFSET_REQ 1

#define PTP_THROW_AWAY_SYNC_OUTLIERS 1

/******** ENDPOINT AUDIO AND CLOCKING PARAMETERS ************************************************/

/* Talker configuration */
#if AVB_DEMO_ENABLE_TALKER

/** The total number of AVB sources (streams that are to be transmitted). */
#define AVB_NUM_SOURCES 1
/** The total number or Talker components (typically the number of
  * tasks running the  :c:func:`avb_1722_talker` function). */
#define AVB_NUM_TALKER_UNITS 1
/** The total number of media inputs (typically number of I2S input channels). */
#define AVB_NUM_MEDIA_INPUTS AVB_DEMO_NUM_CHANNELS
/** Enable the 1722.1 Talker functionality */
#define AVB_1722_1_TALKER_ENABLED 1

#else

#define AVB_NUM_SOURCES 0
#define AVB_NUM_TALKER_UNITS 0
#define AVB_NUM_MEDIA_INPUTS 0
#define AVB_1722_1_TALKER_ENABLED 0
#define AVB_NUM_AUDIO_SDATA_IN 0
#endif

#ifndef AVB_NUM_AUDIO_SDATA_IN
#define AVB_NUM_AUDIO_SDATA_IN ((AVB_NUM_MEDIA_INPUTS+AVB_AUDIO_IF_SAMPLES_PER_PERIOD-1)/AVB_AUDIO_IF_SAMPLES_PER_PERIOD)
#endif

/* Listener configuration */
#if AVB_DEMO_ENABLE_LISTENER

/** The total number of AVB sinks (incoming streams that can be listened to) */
#define AVB_NUM_SINKS 1
/** The total number or listener components
  * (typically the number of tasks running the  :c:func:`avb_1722_listener` function) */
#define AVB_NUM_LISTENER_UNITS 1
/** The total number of media outputs (typically the number of I2S output channels). */
#define AVB_NUM_MEDIA_OUTPUTS AVB_DEMO_NUM_CHANNELS
/** Enable the 1722.1 Listener functionality */
#define AVB_1722_1_LISTENER_ENABLED 1

#else

#define AVB_NUM_SINKS 0
#define AVB_NUM_LISTENER_UNITS 0
#define AVB_NUM_MEDIA_OUTPUTS 0
#define AVB_1722_1_LISTENER_ENABLED 0
#define AVB_NUM_AUDIO_SDATA_OUT 0
#endif

#ifndef AVB_NUM_AUDIO_SDATA_OUT
#define AVB_NUM_AUDIO_SDATA_OUT ((AVB_NUM_MEDIA_OUTPUTS+AVB_AUDIO_IF_SAMPLES_PER_PERIOD-1)/AVB_AUDIO_IF_SAMPLES_PER_PERIOD)
#endif

/** The maximum number of channels permitted per 1722 Talker stream */
#define AVB_MAX_CHANNELS_PER_TALKER_STREAM AVB_DEMO_NUM_CHANNELS
/** The maximum number of channels permitted per 1722 Listener stream */
#define AVB_MAX_CHANNELS_PER_LISTENER_STREAM AVB_DEMO_NUM_CHANNELS

/** Enable combination of the media clock server and PTP server in a single core */
#define COMBINE_MEDIA_CLOCK_AND_PTP 1

/** Use 61883-6 audio format for 1722 streams */
#define AVB_1722_FORMAT_61883_6 1

/** The number of components in the endpoint that will register and initialize media FIFOs
    (typically an audio interface component such as I2S). */
#define AVB_NUM_MEDIA_UNITS 1

/** The number of media clocks in the endpoint. Typically the number of clock domains, each with a
  * separate PLL and master clock. */
#define AVB_NUM_MEDIA_CLOCKS 1

/** The maximum sample rate in Hz of audio that is to be input or output */
#define AVB_MAX_AUDIO_SAMPLE_RATE 96000

// This is the number of master clocks in a word clock
#define MASTER_TO_WORDCLOCK_RATIO 512


/******** 1722.1 PARAMETERS *****************************************************************/

/** Enable 1722.1 AVDECC on the entity */
#define AVB_ENABLE_1722_1 1
#define AVB_1722_1_FIRMWARE_UPGRADE_ENABLED 1

#define AVB_1722_1_ADP_ENTITY_CAPABILITIES (AVB_1722_1_ADP_ENTITY_CAPABILITIES_AEM_SUPPORTED| \
                                            AVB_1722_1_ADP_ENTITY_CAPABILITIES_CLASS_A_SUPPORTED| \
                                            AVB_1722_1_ADP_ENTITY_CAPABILITIES_GPTP_SUPPORTED| \
                                            AVB_1722_1_ADP_ENTITY_CAPABILITIES_EFU_MODE| \
                                            AVB_1722_1_ADP_ENTITY_CAPABILITIES_ADDRESS_ACCESS_SUPPORTED| \
                                            AVB_1722_1_ADP_ENTITY_CAPABILITIES_AEM_IDENTIFY_CONTROL_INDEX_VALID)

#define AVB_1722_1_ADP_MODEL_ID 0x1234

enum aem_control_indices {
    DESCRIPTOR_INDEX_CONTROL_IDENTIFY = 0,
};

/** Enable 1722.1 Controller functionality on the entity. */
#define AVB_1722_1_CONTROLLER_ENABLED 0


/** Enable SRP auto-start and auto-stop a stream when Listeners come and go */
#define SRP_AUTO_TALKER_STREAM_CONTROL 1

#endif


app_avb_lc_demo/src/app_config.h
#ifndef __app_config_h__
#define __app_config_h__

/***** APPLICATION CONFIGURATION ******/

/** Global switch to enable or disable AVB Talker functionality in the demo */
#define AVB_DEMO_ENABLE_TALKER 1
/** Global switch to enable or disable AVB Listener functionality in the demo */
#define AVB_DEMO_ENABLE_LISTENER 1

/** Number of input/output audio channels in the demo application
  * For simplicity, input and output is identical in size but can be configured
  * differently in ``avb_conf.h``. */
#ifndef AVB_DEMO_NUM_CHANNELS
#define AVB_DEMO_NUM_CHANNELS 8
#endif

/***** PORTS *********/

// This include fill will get included if you use a XMOS development target
// e.g. the XR-AVB-LC-BRD avb kit or the XMOS sliceKIT and set
// all the port defines accordingly

#ifdef __avb_app_board_config_h_exists__
#include "avb_app_board_config.h"
#endif

// If you are not using a standard dev board (e.g. creating an application for
// your own board), you can update the port mappings here



#define I2S_CLK_1 XS1_CLKBLK_3
#define I2S_CLK_2 XS1_CLKBLK_4

#if !USING_XMOS_DEV_BOARD

#define ETHERNET_PHY_ADDRESS 0

#define PORT_ETH_RXCLK     on tile[1]: XS1_PORT_1A
#define PORT_ETH_ERR       on tile[1]: XS1_PORT_1B
#define PORT_ETH_RXD       on tile[1]  XS1_PORT_4C
#define PORT_ETH_RXDV      on tile[1]: XS1_PORT_1D
#define PORT_ETH_TXCLK     on tile[1]: XS1_PORT_1C
#define PORT_ETH_TXEN      on tile[1]: XS1_PORT_1E
#define PORT_ETH_TXD       on tile[1]: XS1_PORT_4D

// ETHERNET_PORTS
#define ETHERNET_MII_INIT_full \
   { ETHERNET_CLKBLK_0, ETHERNET_CLKBLK_1, \
     PORT_ETH_RXCLK, PORT_ETH_ERR, PORT_ETH_RXD, PORT_ETH_RXDV, \
     PORT_ETH_TXCLK, PORT_ETH_TXEN, PORT_ETH_TXD }

#if SMI_COMBINE_MDC_MDIO
#define PORT_ETH_MDIOC
#define ETHERNET_SMI_INIT {ETHERNET_PHY_ADDRESS, \
                           PORT_ETH_MDIOC}
#else
#define PORT_ETH_MDIO     on tile[1]:XS1_PORT_1G
#define PORT_ETH_MDC      on tile[1]:XS1_PORT_1G
#define ETHERNET_SMI_INIT {ETHERNET_PHY_ADDRESS, \
                           PORT_ETH_MDIO,                \
                           PORT_ETH_MDC}
#endif

// CODEC control ports
#define PORT_I2C_SCL      on tile[1]:XS1_PORT_1L
#define PORT_I2C_SDA      on tile[1]:XS1_PORT_1K

// PLL frequency control
#define PORT_SYNC_OUT     on tile[0]:XS1_PORT_1E

// I2S ports
#define PORT_MCLK         on tile[0]:XS1_PORT_1K
#define PORT_SCLK         on tile[0]:XS1_PORT_1F
#define PORT_LRCLK        on tile[0]:XS1_PORT_1L

#define USER_PORT_SDATA_OUT    {on tile[0]:XS1_PORT_1G,on tile[0]:XS1_PORT_1H,on tile[0]:XS1_PORT_1I,on tile[0]:XS1_PORT_1J}

#define PORT_SDATA_OUT       USER_PORT_SDATA_OUT

#define USER_PORT_SDATA_IN     {on tile[0]:XS1_PORT_1M,on tile[0]:XS1_PORT_1N,on tile[0]:XS1_PORT_1O,on tile[0]:XS1_PORT_1P}

#define PORT_SDATA_IN       USER_PORT_SDATA_IN

#define USER_PORT_BUTTONS     on tile[1]:XS1_PORT_4A
#define PORT_BUTTONS USER_PORT_BUTTONS

#define USER_PORT_MUTE_LED_REMOTE  on tile[1]:XS1_PORT_4E
#define PORT_MUTE_LED_REMOTE USER_PORT_MUTE_LED_REMOTE

#define USER_PORT_LEDS  on tile[1]:XS1_PORT_4F
#define PORT_LEDS USER_PORT_LEDS

#endif // !USING_XMOS_DEV_BOARD




#endif // __app_config_h__


module_avb_audio/src/xstream_subset/xports-i2s.h
#ifndef XPORTS_I2S
#define XPORTS_I2S

#include "xstream-types.h"
#include "xstream-options.h"
#include "xstream-settings.h"

// ====================================================================================================================
// Driver Overview
// ====================================================================================================================

// This driver supports most combininatons of sampling frequency, slot counts and sizes, word sizes and formmating, and
// ADC/DAC wire counts ...
//
// Slot Count:     Supports 2, 4, 6, 8, 16, and 32 words per audio cycle.  A slot count of two is standard two-word (left
//                 and right) stereo PCM/I2S.
//
// Slot Size:      Supports 16, 24 and 32 bit-clock cycles per sample word slot.
//
// Word Size:      Supports 16, 24 and 32-bit signed 2's compliment sample words.
//
// Frame Sync:     Supports PCM left/right clock (aligned to sample word), I2S left/right clock (aligned with sample word
//                 minus one bit-clock cycle), and TDM frame-sync pulse.
//
// Word Format:    Supports left-justified MSB-first and right-justified MSB-first sample word alignment.
//
// Sampleing Rate: Supports single (44100 KHz, 48000 KHz), double (88200 KHz, 96000 KHz), quad (176400 KHz, 192000 KHz),
//                 and octave (352800 KHz, 384000 KHz) sampling rates.
//
// Master Clock:   Supports internaly generated MCLK (derived from the XS1 system clock) or an external MCLK.
////
// Most combininatons of sampling frequency, slot counts and sizes, word sizes and formmating, and ADC/DAC wire counts
// are supported.  Unsupported combinations are detected at compile-time (via checking of the preprocessor #defines used
// to define the configration).  Invalid combinitations are either due to either the lack of implementation (not done yet)
// or to the innability of the XS1 core to meet timing/performance requirements.
//
// Valid slot count and slot size combininations (slot count / slot size): 2/16 2/32 4/16 4/24 4/32 6/16 6/32
//                                                                         8/16 8/24 8/32 16/16 16/24 16/32
//                                                                         32/16 32/24 32/32
//
// Valid slot size and word size comibinations (slot size / word size): 32/32 32/24 32/16 24/24 24/16 16/16
//
// Valid sample frequency, port width, slot size and DAC/ADC wire count combinations:
//
//    Sample Freq   Port Width   Slot Size   DAC/ADC Wires
//    single        1            16          1/2/3/4
//    single        1            24          1/2/3/4
//    single        1            32          1/2/3/4
//    double        1            16          1/2/3/4
//    double        1            24          1/2/3
//    double        1            32          1/2/3/4
//    quad          1            16          1/2/3/4
//    quad          1            24          1/2
//    quad          1            32          1/2/3/4
//    octave        1            16          1/2/3/4
//    octave        1            24          1
//    octave        1            32          1/2/3/4

// ====================================================================================================================
// Driver Instance Context
// ====================================================================================================================

typedef struct
{
	clock                clock_mclk;
	clock                clock_bclk;

	#if XPORTS_I2S__MASTER_CLOCK_SOURCE == XPORTS_I2S__MASTER_CLOCK_SOURCE__EXTERNAL
	in port              pin_mclk;
	#elif XPORTS_I2S__MASTER_CLOCK_SOURCE == XPORTS_I2S__MASTER_CLOCK_SOURCE__INTERNAL
	out port             pin_mclk;
	#endif

	#if XPORTS_I2S__BUS_CLOCKING_MODE == XPORTS_I2S__BUS_CLOCKING_MODE__MASTER
	out buffered port:32 pin_bclk;
	out buffered port:32 pin_wclk;
	#elif XPORTS_I2S__BUS_CLOCKING_MODE == XPORTS_I2S__BUS_CLOCKING_MODE__SLAVE
	in buffered port:32  pin_bclk;
	in buffered port:32  pin_wclk;
	#endif

	#if XPORTS_I2S__DAC_DATA_WIRE_COUNT > 0
	out buffered port:32 pin_dac[XPORTS_I2S__DAC_DATA_WIRE_COUNT];
	#endif

	#if XPORTS_I2S__ADC_DATA_WIRE_COUNT > 0
	in  buffered port:32 pin_adc[XPORTS_I2S__ADC_DATA_WIRE_COUNT];
	#endif

	unsigned int first_pass;
	unsigned int mclk_bclk_ratio;
}
xports_i2s__context_t;

// ====================================================================================================================
// Driver Public Interface
// ====================================================================================================================

// Should be performed before any other API function call.

void xports_i2s__initialize( xports_i2s__context_t& self );

// To be called to when MCLK or the sampling frequency need to change.

void xports_i2s__configure ( xports_i2s__context_t& self, unsigned int mclk_frequency, unsigned int sample_rate );

// Pack sample data into byte stream approrpriate for DAC output

#if XPORTS_I2S__DATA_FORMAT_MODE == XPORTS_I2S__DATA_FORMAT_MODE__EXTERNAL

	void xports_i2s__pack  ( xports_i2s__context_t& self, unsigned int buffer_address,  unsigned int samples_address );

#endif

// Perform sample data transfer (one sample for each PCM/I2S left/right or TDM slot) for one audio cycle

void xports_i2s__transfer  ( xports_i2s__context_t& self, unsigned int dac_buffer_address, unsigned int adc_buffer_address );

void xports_i2s__transfer_chan( xports_i2s__context_t& self, streaming chanend c_dac_samples, streaming chanend c_adc_samples );

// Unpack sample data from byte stream read from ADC input

#if XPORTS_I2S__DATA_FORMAT_MODE == XPORTS_I2S__DATA_FORMAT_MODE__EXTERNAL

	void xports_i2s__unpack( xports_i2s__context_t& self, unsigned int samples_address, unsigned int buffer_address  );

#endif

// ====================================================================================================================
// ====================================================================================================================

#endif


module_avb_audio/src/xstream_subset/xstream-settings.h
#include "xstream-options.h"
#include <avb_conf.h>

#define XSTREAME_SHARED_MEM_IF 0

#define XDEVICE_CODE__CS4270_I2C_ADDRESS    XDEVICE_CODEC__CS4270_I2C_ADDRESS_1

#define XPORTS_I2S__ADC_DATA_WIRE_COUNT     AVB_NUM_AUDIO_SDATA_IN
// Don't do this in unless XSTREAME_SHARED_MEM_IF: #define XPORTS_I2S__ADC_DATA_WIRE_COUNT     1
#define XPORTS_I2S__ADC_WORD_FORMAT         XPORTS_I2S__WORD_FORMAT__LEFT_JUST
#define XPORTS_I2S__ADC_WORD_SIZE           XPORTS_I2S__WORD_SIZE__32_BIT
#define XPORTS_I2S__BUS_CLOCKING_MODE       XPORTS_I2S__BUS_CLOCKING_MODE__MASTER
#define XPORTS_I2S__DAC_DATA_WIRE_COUNT     AVB_NUM_AUDIO_SDATA_OUT
// Don't do this in unless XSTREAME_SHARED_MEM_IF:#define XPORTS_I2S__DAC_DATA_WIRE_COUNT     1
#define XPORTS_I2S__DAC_WORD_FORMAT         XPORTS_I2S__WORD_FORMAT__LEFT_JUST
#define XPORTS_I2S__DAC_WORD_SIZE           XPORTS_I2S__WORD_SIZE__32_BIT
#define XPORTS_I2S__DATA_FORMAT_MODE        XPORTS_I2S__DATA_FORMAT_MODE__EXTERNAL

#define XPORTS_I2S__DATA_LOOPBACK_MODE      XPORTS_I2S__DATA_LOOPBACK_MODE__OFF

#define XPORTS_I2S__MASTER_CLOCK_SOURCE     XPORTS_I2S__MASTER_CLOCK_SOURCE__EXTERNAL
#define XPORTS_I2S__MCLK_BCLK_RATIO         XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_2
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER    XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X6
#define XPORTS_I2S__OPTIMIZATION_MODE       XPORTS_I2S__OPTIMIZATION_MODE__SPEED
#define XPORTS_I2S__SLOT_COUNT              AVB_AUDIO_IF_SAMPLES_PER_PERIOD
#define XPORTS_I2S__SLOT_SIZE               XPORTS_I2S__SLOT_SIZE__32_BIT
#define XPORTS_I2S__SYNC_WORD_FORMAT        XPORTS_I2S__SYNC_WORD_FORMAT__TDM_STYLE


module_avb_audio/src/xstream_subset/xstream-options.h

// ====================================================================================================================
// Configurations for all modules in the XSTREAM library
// ====================================================================================================================

#ifndef XSTREAM_OPTIONS
#define XSTREAM_OPTIONS

// ====================================================================================================================
// XPORTS Protocol Options for Each I2S Instance
// ====================================================================================================================

#define XPORTS_I2S__OPTIMIZATION_MODE__SPEED       0
#define XPORTS_I2S__OPTIMIZATION_MODE__SIZE        1

#define XPORTS_I2S__DATA_LOOPBACK_MODE__OFF        0  // Data is not looped back
#define XPORTS_I2S__DATA_LOOPBACK_MODE__ON         1  // DAC data is looped back to ADC data (ADC sampled values are overwritten)

#define XPORTS_I2S__DATA_FORMAT_MODE__EXTERNAL     0  // Perform DAC and ADC sample data rendering/parsing in application
#define XPORTS_I2S__DATA_FORMAT_MODE__INTERNAL     1  // Perform DAC and ADC sample data rendering/parsing in codec module

#define XPORTS_I2S__MASTER_CLOCK_SOURCE__EXTERNAL  0  // Use external MCLK signal (MCLK pin is an input)
#define XPORTS_I2S__MASTER_CLOCK_SOURCE__INTERNAL  1  // Generate MCLK from internal oscillator (MCLK pin in an output)

#define XPORTS_I2S__BUS_CLOCKING_MODE__MASTER      0  // Generate BCLK and WCLK signals (outputs)
#define XPORTS_I2S__BUS_CLOCKING_MODE__SLAVE       1  // Sample BCLK and WCLK signals (inputs)

#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_1        1  // MCLK frequency / BCLK frequency = 1.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_2        2  // MCLK frequency / BCLK frequency = 2.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_4        4  // MCLK frequency / BCLK frequency = 4.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_8        8  // MCLK frequency / BCLK frequency = 8.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_16      16  // MCLK frequency / BCLK frequency = 16.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__ONLY_32      32  // MCLK frequency / BCLK frequency = 32.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__1_TO_32      (32+16+8+4+2+1)  // (MCLK frequency / BCLK frequency) can be 1.0 or 2.0 ... or 32.0
#define XPORTS_I2S__MCLK_BCLK_RATIO__2_TO_32      (32+16+8+4+2)    // (MCLK frequency / BCLK frequency) can be 2.0 ... or 32.0

// FS Base @ BitsPerSample                              // 48K0 @ 32/16,  44K1 @ 32/16,     48K0 @ 24,     44K1 @ 24
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X6       1  // 49152000 Mhz,  45158400 Mhz,  36864000 Mhz,  33868800 Mhz
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X5       2  // 24576000 Mhz,  22579200 Mhz,  18432000 Mhz,  16934400 Mhz
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X4       4  // 12288000 Mhz,  11289600 Mhz,   9216000 Mhz,   8467200 Mhz
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X3       8  //  6144000 Mhz,   5644800 Mhz,   4608000 Mhz,   4233600 Mhz
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X2      16  //  3072000 Mhz,   2822400 Mhz,   2304000 Mhz,   2116800 Mhz
#define XPORTS_I2S__MCLK_FREQ_MULTIPLIER__X1      32  //  1536000 Mhz,   1411200 Mhz,   1152000 Mhz,   1058400 Mhz

#define XPORTS_I2S__SLOT_SIZE__16_BIT             16  // 16 BCLK's per ADC/DAC sample, contains one 16-bit sample word
#define XPORTS_I2S__SLOT_SIZE__24_BIT             24  // 24 BCLK's per ADC/DAC sample, contains one 16/24-bit sample word
#define XPORTS_I2S__SLOT_SIZE__32_BIT             32  // 32 BCLK's per ADC/DAC sample, contains one 16/24/32-bit sample word

#define XPORTS_I2S__SLOT_COUNT__2_PER_CYCLE        2  // 2 slots - 2 samples per audio cycle (a.k.a stereo)
#define XPORTS_I2S__SLOT_COUNT__4_PER_CYCLE        4  // 4 slots - 4 samples per audio cycle (TDM I4S)
#define XPORTS_I2S__SLOT_COUNT__6_PER_CYCLE        4  // 6 slots - 6 samples per audio cycle (TDM I6S)
#define XPORTS_I2S__SLOT_COUNT__8_PER_CYCLE        8  // 8 slots - 8 samples per audio cycle (TDM I8S)
#define XPORTS_I2S__SLOT_COUNT__16_PER_CYCLE      16  // 16 slots - 16 samples per audio cycle (TDM I16S)
#define XPORTS_I2S__SLOT_COUNT__32_PER_CYCLE      32  // 32 slots - 32 samples per audio cycle (TDM I32S)

#define XPORTS_I2S__SYNC_WORD_FORMAT__PCM_STYLE    0  // Left=High/Right=Low word-clock formatting on WCLK/LRCLK line
#define XPORTS_I2S__SYNC_WORD_FORMAT__I2S_STYLE    1  // Left=Low/Right=High (1 bit-clock skew left) word-clock formatting on WCLK/LRCLK line
#define XPORTS_I2S__SYNC_WORD_FORMAT__TDM_STYLE    2  // TDM sync pulse (1 bit-clock skew left) per audio cycle on WCLK/LRCLK line

#define XPORTS_I2S__WORD_SIZE__16_BIT             16  // 16-bit sample word size, padded with zero's to slot size
#define XPORTS_I2S__WORD_SIZE__24_BIT             24  // 24-bit sample word size, padded with zero's to slot size
#define XPORTS_I2S__WORD_SIZE__32_BIT             32  // 32-bit sample word size, padded with zero's to slot size

#define XPORTS_I2S__WORD_FORMAT__LEFT_JUST         0  // Left justfied sample word, justified to left side of slot word
#define XPORTS_I2S__WORD_FORMAT__RIGHT_JUST        1  // Right justified sample word, justified to right side of slot word

#define XPORTS_I2S__DATA_WIRE_COUNT__0             0  // No data wire for PCM/I2S bus (doesn't use a port)
#define XPORTS_I2S__DATA_WIRE_COUNT__1             1  // One data wire for PCM/I2S bus (uses one 1-bit port)
#define XPORTS_I2S__DATA_WIRE_COUNT__2             2  // Two data wires for PCM/I2S bus (uses two 1-bit ports)
#define XPORTS_I2S__DATA_WIRE_COUNT__3             3  // Three data wires for PCM/I2S bus (uses three 1-bit ports)
#define XPORTS_I2S__DATA_WIRE_COUNT__4             4  // Four data wires for PCM/I2S bus (uses a one 4-bit port or four 1-bit ports)

// ====================================================================================================================
// XPORTS Protocol Options for Each I2C Instance
// ====================================================================================================================

#define XPORTS_I2C__OPTIMIZATION_MODE__SPEED       0
#define XPORTS_I2C__OPTIMIZATION_MODE__SIZE        1

#define XPORTS_I2C__DATA_LOOPBACK_MODE__OFF        0  // Data is not looped back
#define XPORTS_I2C__DATA_LOOPBACK_MODE__ON         1  // DAC data is looped back to ADC data (ADC sampled values are overwritten)

#define XPORTS_I2C__BUS_CLOCKING_MODE__MASTER      0  // Generate BCLK and WCLK signals (outputs)
#define XPORTS_I2C__BUS_CLOCKING_MODE__SLAVE       1  // Sample BCLK and WCLK signals (inputs)

#define XPORTS_I2C__DATA_BIT_RATE__CUSTOM          0
#define XPORTS_I2C__DATA_BIT_RATE__100_KBPS        1
#define XPORTS_I2C__DATA_BIT_RATE__400_KBPS        2

#define XPORTS_I2C__RESTART_CONDITIONS__DISABLED   0
#define XPORTS_I2C__RESTART_CONDITIONS__ENABLED    1

#define XPORTS_I2C__CLOCK_STRETCHING__DISABLED     0
#define XPORTS_I2C__CLOCK_STRETCHING__ENABLED      1

#define XPORTS_I2C__COLLISION_DETECTION__DISABLED  0
#define XPORTS_I2C__COLLISION_DETECTION__ENABLED   1

// ====================================================================================================================
// XDEVICE Protocol Options for Each CODEC Instance
// ====================================================================================================================

#define XDEVICE_CODEC__CS4270_I2C_ADDRESS_1  0b01001100
#define XDEVICE_CODEC__CS4270_I2C_ADDRESS_2  0b01001101
#define XDEVICE_CODEC__CS4270_I2C_ADDRESS_3  0b01001110
#define XDEVICE_CODEC__CS4270_I2C_ADDRESS_4  0b01001111

// ====================================================================================================================
// XDEVICE Protocol Options for Each RADIO Instance
// ====================================================================================================================

// ====================================================================================================================
// XAUDIO Instance Assignments
// ====================================================================================================================

#define XIO_AUDIO__INTERFACE_STANDARD        1
#define XIO_AUDIO__INTERFACE_UAC2            2

#define XIO_AUDIO__OUTPUT_CHANNEL_COUNT_2    2
#define XIO_AUDIO__OUTPUT_CHANNEL_COUNT_32   32

#define XIO_AUDIO__INPUT_CHANNEL_COUNT_2     2
#define XIO_AUDIO__INPUT_CHANNEL_COUNT_32    32

// ====================================================================================================================
// ====================================================================================================================

#endif
